### ISSUE TITLE
More ergonomic way to pass multiple columns to `select`/`with_columns`/`groupby`/... 

### ISSUE THREAD

**Issue by stinodego** [pola-rs/polars#6451]
### Problem description

TL;DR: I would like to be able to write `df.select("a", "b")` instead of `df.select(["a", "b"])`

### The problem

We have multiple often-used methods that can take either a single string/expression, or a list of multiple strings/expressions. Specifying multiple expressions must be done by specifying passing a list. **This can be a bit cumbersome**:

* Since Polars expressions can get pretty long, multiple columns will often live on separate lines. Having to specify a list introduces an extra level of indentation _(see the README for a great example of this - for a long time, you had to sidescroll to fully read it)_:

```python
df.select(
    [
        pl.col("a").filter(...),  # long expression
        pl.col("b").filter(...),  # long expression
    ]
)
```

* If I want to select an extra column in my statement `df.select("a")`, I have to add brackets first and write `df.select(["a", "b"])`. And if I decide not to select `"b"` after all, I have to remove the brackets again  _(ok, that last part technically not required, but a single item list does not make much sense here)_ .

### We can do better

By leveraging the `*args` parameter, we can allow users to write:

```python
df.select(
    pl.col("a").filter(...),  # long expression
    pl.col("b").filter(...),  # long expression
)
```

**Less fiddling with brackets, less indentation!**

The function signature would look something like:

```python
def select(
    self,
    *exprs: str | pli.Expr,
)
```

`*exprs` will contain a tuple of columns specified. **We can add this in a non-breaking way** by checking if a single list was passed and parsing accordingly. Later we may decide to remove the list syntax altogether; that will make parsing the arguments even easier than it is now!

Some methods take additional parameters besides a (list of) expressions. These can still facilitate the same ergonomics, but require that the other parameters are keyword-only arguments. **This requires a breaking change.** For example, `groupby` would look like this:

```python
def groupby(
    self,
    *by: str | pli.Expr,
    maintain_order: bool = False,
)
```

And calling it would look like:
```python
df.groupby("a", "b").agg(...)  # new ergonomic syntax
df.groupby(["a", "b"]).agg(...)  # this still works
df.groupby(["a", "b"], "c").agg(...)  # we will throw an exception in this case
df.groupby("a", "b", maintain_order=True).agg(...)  # maintain_order must be a keyword argument
df.groupby("a", True).agg(...)  # this old syntax is now BROKEN
```

### The methods affected by this change

We can introduce this for any method that takes either a single or multiple expressions. I have identified the following candidates:

* [x] `select`
* [x] `with_columns`
* [x] `groupby`
* [x] `agg`
* [x] `over`
* [x] `sort`
* [x] `sort_by`
* [x] `coalesce`
* [x] `col`
* [x] `drop`
* [x] `exclude`
* [x] `partition_by`
* [x] `explode`
* [x] `unnest`
* [x] `max`
* [x] `min`
* [x] `sum`
* [x] `cumsum`
* [x] `any`
* [x] `all`
* [x] `concat_list`

The following methods should also be adjusted, but will require a breaking change due to containing keyword arguments that are not currently keyword-only:
* [x] `struct`
* [x] `arg_sort_by`
* [x] `concat_str`

There might be others; I will add them here if I encounter them.

### Implementation

1. The first step would be to identify which methods qualify.
2. The second step would be to make the breaking change for `groupby` and similar methods to make some parameters keyword-only (maybe this can even happen in `0.16.0` still?).
3. The third step would be to introduce this non-breaking ergonomic change across all methods.

### Trade-offs

Pro:
* More ergonomic user experience
* Easier parsing of arguments

Con:
* ??

I don't really see any negatives here, but I'm curious about your thoughts!

For reference, [pyspark already does this](https://spark.apache.org/docs/3.1.1/api/python/reference/api/pyspark.sql.DataFrame.select.html) for some of their methods.

**Comment by universalmind303** [comment]
+1 to this. 

The node bindings take variadic functions for select, groupby, ...

**Comment by ritchie46** [comment]
I agree. As long as we both accept `*args` and a list. 

I also think that in those elegible methods. All non-expressions should be keyword only.

**Comment by alexander-beedie** [comment]
I wouldn't necessarily rush this for `0.16.0` though; I implemented several syntax tokens using this style and you'd be amazed how many people pass-in lists (often generated by other code) and forget to `*splat` them into the method until they internalise the pattern - it happens a _lot_ ;) 

Being able to handle `*args` _or_ a list helps, but if the function could accept (for example) a list of lists then it's not possible to determine which one was intended, and you have to come down on one side or the other. A good look over the potential target functions is required here. If we plan to move to only one supported pattern (eg: only `*args` ) then at least that could form a basis for evaluating ambiguous calls (_if_ we have such functions - if we don't, it gets a lot easier!)

**Comment by stinodego** [comment]
There's definitely some edge cases to be worked out, as you point out.

The only thing I'd like to get into the next breaking release is making the non-expression arguments keyword-only. I think this is quite minor as I think it's natural to pass these as keywords already. This is something that Ritchie also supports as he states above.

Once we have that in place, we can take our sweet time looking into adding the new syntax in a non-breaking way.

**Comment by JulianCologne** [comment]
I understand the intention of this issue, but I think this favors ergonomics (faster/shorter code) over explicitness (clearer/readable code).
For me personally, I always prefer readability and explicitness over brevity and ergonomics!

What this means in this context:

`df.select("a", "b")`
- very short, but:
    - what am I selecting? rows/columns/individual elements?
    - not clear what the arguments are, "Do I only get 'a'? What about the 'b' parameter?"
    - many people are not familiar with `*args`

`df.select(columns=["a", "b"])`
- more verbose, but:
    - clear what is happening and what the arguments are (no docs required, a great API is self-documenting)
- **please change `exprs` argument to `columns`. Expressions are used everywhere in Polars and it is not clear what `exprs` means here, especially because it accepts so many different types that are not expressions!**

Especially for new users and beginners, I think it is important to be as explicit as possible.

Also for methods with more than one argument I much prefer the explicit version.
```python
# short
df.groupby(
    "a",
    "b",
    maintain_order=True,
)
df.groupby("a", "b", maintain_order=True)

# explicit
df.groupby(  # would also love to rename `groupby` to `group` to avoid the double `by`
    by=["a", "b"],
    maintain_order=True,
)
```

Since `*args` prevents the use of the keyword argument (explicitness/readability), this is a downvote from me.

**Comment by ghuls** [comment]
Keywords only arguments would also be great to have for e.g. read_csv or other functions with a lot of parameters, so in case it is needed in the future, the argument names can be changed or rearranged without silently breaking existing code for people using positional arguments for those functions.

**Comment by stinodego** [comment]
> **please change `exprs` argument to `columns`.**

I think that would be a sensible change. I don't like `exprs` myself.

We were actually thinking of making some arguments positional only (#5960), but I have to say that you make a compelling case against that.

**Comment by alexander-beedie** [comment]
> The only thing I'd like to get into the next breaking release is making the non-expression arguments keyword-only. --//-- Once we have that in place, we can take our sweet time looking into adding the new syntax in a non-breaking way.

Sounds good! :)

**Comment by zundertj** [comment]
I am not so convinced yet on the change from `exprs` to `columns`, because it documents the output (i.e. the dataframe returned has those columns), but not the inputs anymore. `columns` reads to me as existing (string) column names, which suggests that you can only pass in strings? But maybe that is just the association I am making. I do agree `exprs` is a terrible name to use for a keyword argument, but in code, it is almost always clearer to call it as an arg, not as kwarg. Writing `df.select(columns=[...])` becomes very annoying quickly, and I think that although it is a small hurdle, this is not an insurmountable one?

Making parameters keyword only is also a sensible step, but it does get you tangled up with the experimental feature `pl.Config.with_columns_kwargs=True`
```python
df.with_columns(pl.lit(a)*2, by=pl.col("c"))
```
In this example it is clear for experienced users there are no "real" kwargs here, but for novices they might assume that `by` is somehow special. I would favor keyword only over this smarter syntax to be honest, although I like this syntax as well (better than writing `.alias()` at the end of your expression every time).

**Comment by alexander-beedie** [comment]
FYI: `with_columns` kwargs are losing their "experimental" tag with https://github.com/pola-rs/polars/pull/6497 (and gaining auto-structification for multi-output expressions) 👍

**Comment by cmdlineluser** [comment]
I see `pl.struct` is on the consideration list.

Do you think it could also be a candidate for `**named_exprs`?

e.g.
```
pl.struct(
   one = pl.col(...),
   two = pl.col(...),
   three = pl.col(...)
)
```
Instead of
```
pl.struct([
   pl.col(...).alias("one"),
   pl.col(...).alias("two"),
   pl.col(...).alias("three")
])
```

**Comment by stinodego** [comment]
Seems like a good idea to me. I also added that syntax to `groupby.agg(...)`. But this will have to wait until the next breaking release, because the keyword arguments of `struct` are not keyword-only yet.

**Comment by stinodego** [comment]
Yes, I actually ran into this one myself last week! Will add those to the list. Thanks for the suggestion.

**Comment by bordumb** [comment]
This is a good idea.

FWIW: it might be worth looking into PySpark as it implements this same `groupBy("a", "b", ...)` syntax
